### PR TITLE
Platml 2828

### PR DESCRIPTION
**Pull request by tumulurik**

### CODE REVIEW

**Review by sathyarun18 — APPROVED** [review]

File: recipes/pyspark/README.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                     "isHoliday": {
+                       "id": "isHoliday",
+                       "type": "boolean"
+                     }
+                   }
+             }
+### Upload Dataset from UI             
+
+Upload the parquet file from here:
+`https://github.com/adobe/experience-platform-dsw-reference/blob/master/datasets/retail/DSWRetailSales.parquet`
```
Comment (rjrussell77) [inline recipes/pyspark/README.md:87]: What version is this based on (XDM 0.9.9.?)  Does it need updating?

File: recipes/pyspark/pysparkretailapp/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    assembler = VectorAssembler(inputCols=cols, outputCol="features")
+
+    # Construct the pipeline
+    pipeline = Pipeline(stages=[assembler, gbt])
+
+    return pipeline
+
+
+def getParamMap(configProperties, sparkSession):
+    return None
```
Comment (rjrussell77) [inline recipes/pyspark/pysparkretailapp/pipeline.py:56]: This is curious.

File: recipes/pyspark/pysparkretailapp/scoringdataloader.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        raise ValueError("spark parameter is null")
+
+    service_token = str(spark.sparkContext.getConf().get("ML_FRAMEWORK_IMS_ML_TOKEN"))
+    user_token = str(spark.sparkContext.getConf().get("ML_FRAMEWORK_IMS_TOKEN"))
+    org_id = str(spark.sparkContext.getConf().get("ML_FRAMEWORK_IMS_ORG_ID"))
+
+    dataset_id = str(configProperties.get("scoringDataSetId"))
+    api_key = str(configProperties.get("api_key"))
+
+    for arg in ['service_token', 'user_token', 'org_id', 'dataset_id', 'api_key']:
```
Comment (rjrussell77) [inline recipes/pyspark/pysparkretailapp/scoringdataloader.py]: Seeing repeated instances of same literal strings.  Better to make a constant?

**Review by rjrussell77 — APPROVED** [review]
LGTM w/ minor comments

**Review by tatsuyawi — APPROVED** [review]

File: recipes/pyspark/pysparkretailapp/scoringdataloader.py (hunk 3 above)
Comment (tumulurik) [reply recipes/pyspark/pysparkretailapp/scoringdataloader.py]: Added a helper

File: recipes/pyspark/README.md (hunk 1 above)
Comment (tumulurik) [reply recipes/pyspark/README.md:87]: Will capture this validation as part of the ticket - https://jira.corp.adobe.com/browse/PLATML-4247

File: recipes/pyspark/pysparkretailapp/pipeline.py (hunk 2 above)
Comment (tumulurik) [reply recipes/pyspark/pysparkretailapp/pipeline.py:56]: Needs to be present but an empty placeholder.

File: recipes/pyspark/pysparkretailapp/trainingdataloader.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# from Adobe.
+#####################################################################
+
+from pyspark.sql.functions import lit
+from .helper import *
+
+
+def load(configProperties, spark):
+
+    if(configProperties is  None) :
```
Comment (rjrussell77) [inline recipes/pyspark/pysparkretailapp/trainingdataloader.py]: Small typo - extra space here: "is   None"

**Review by rjrussell77 — APPROVED** [review]
One small typo but otherwise, LGTM

File: recipes/pyspark/pysparkretailapp/trainingdataloader.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    if configProperties is None :
+        raise ValueError("configProperties parameter is null")
+    if spark is None:
+        raise ValueError("spark parameter is null")
+
+    pd = loaddataset(configProperties, spark, "trainingDataSetId")
+    pd = preparedataset(pd)
+
+    # Split the data
+    train = pd.filter(pd["tx_date"] <= lit('2012-01-27'))
```
Comment (jonaphin) [inline recipes/pyspark/pysparkretailapp/trainingdataloader.py:33]: why this date? lit('2012-01-27')

File: recipes/pyspark/pysparkretailapp/helper.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+
+from pyspark.ml.feature import StringIndexer
+from pyspark.sql.functions import col
+from pyspark.sql.types import IntegerType
+from pyspark.sql.functions import unix_timestamp, from_unixtime, to_date, lit, lag, udf, date_format
+from pyspark.sql import Window
+
+
+def loaddataset(configProperties, spark, taskId):
```
Comment (jonaphin) [inline recipes/pyspark/pysparkretailapp/helper.py]: naming convention of the method is a little ...un-camelCasey :)

File: recipes/pyspark/pysparkretailapp/helper.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    pd = spark.read.format("com.adobe.platform.dataset") \
+        .option('serviceToken', service_token) \
+        .option('userToken', user_token) \
+        .option('orgId', org_id) \
+        .option('serviceApiKey', api_key) \
+        .load(dataset_id)
+    return pd
+
+
+def preparedataset(dataset):
```
Comment (jonaphin) [inline recipes/pyspark/pysparkretailapp/helper.py]: same comment about camel casing

**Review by jonaphin — APPROVED** [review]

File: recipes/pyspark/pysparkretailapp/trainingdataloader.py (hunk 5 above)
Comment (tumulurik) [reply recipes/pyspark/pysparkretailapp/trainingdataloader.py:33]: We do the sorting of the data based on the computed timestamp from the date and store and then split the data.

File: recipes/pyspark/pysparkretailapp/helper.py (hunk 6 above)
Comment (tumulurik) [reply recipes/pyspark/pysparkretailapp/helper.py]: Agree, fixed this.

**Pull request merged**